### PR TITLE
DBZ-965 Track latest LSN per each change

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-965
https://issues.jboss.org/browse/DBZ-970

### CODE REVIEW

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java [hunk 1]
```diff
@@ -58,6 +58,7 @@
     private final AtomicReference<ReplicationStream> replicationStream;
     private final AtomicBoolean cleanupExecuted = new AtomicBoolean();
     private PgConnection typeResolverConnection = null;
+    private Long lastProcessedLsn;
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: Same comment about naming as above.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java [hunk 2]
```diff
@@ -270,11 +276,13 @@ private void process(ReplicationMessage message, Long lsn, BlockingConsumer<Chan
             }
         }
 
-        heartbeat.heartbeat(sourceInfo.partition(), sourceInfo.offset(),
-                r -> consumer.accept(new ChangeEvent(r, message.isLastEventForLsn())));
+        if (message.isLastEventForLsn()) {
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: Why is the heartbeat emitted only for the last event of an LSN now?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 3]
```diff
@@ -376,6 +376,11 @@ public MessageDecoder messageDecoder() {
             public MessageDecoder messageDecoder() {
                 return new StreamingWal2JsonMessageDecoder();
             }
+
+            @Override
+            public String getPostgresPluginName() {
```
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java]: Actually pushed a commit for this.

**Comment by jpechane** [comment]
@gunnarmorling Comments incorporated

**Comment by gunnarmorling** [comment]
@jpechane Another thing coming to mind is testing this. It seems it should be doable by setting up the whitelist config as needed, altering two records in a TX with the second one not being whitelisted and then asserting on that PG view for confirmed LSNs by replication slot. I'd prefer to have this test to prevent regressions here, in particular as we'll rewrite the connector to make use of the new framework classes.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java (hunk 2 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: LSN flushes for non-complete LSNs are lost so I thought it is better to keep LSN flushes and offset change sin lock-step.

Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: Ok, makes sense.

**Comment by gunnarmorling** [comment]
@jpechane Pushed one more commit for some minor clean-up. Once we have the test, it's good to go from my POV.

**Comment by jpechane** [comment]
@gunnarmorling The test has been added.

**Comment by gunnarmorling** [comment]
Rebased and applied. Reworked the last commit to avoid a CS failure which it had introduced. Thanks a lot, @jpechane!

**Pull request closed without merging**